File: leetcode/medium/376-wiggle-sub/lin-dp.cpp
/**
 * @file lin-dp.cpp
 * @author Vortexx2
 * @brief Problem 376 - Wiggle Subsequence
 * @approach Using a linear DP approach
 * @date 26-08-2022
 * 
 * Runtime - 5 ms O(n)
 * Memory Usage - 7.3 MB O(n)
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int wiggleMaxLength(vector<int>& nums) {
    int n = nums.size();

    if (n < 2) return nums.size();

    // a particular element shows the longest subsequence with a positive wiggle
    // as the last wiggle
    vector<int> ups(n, 1);
    // a particular element shows the longest subsequence with a negative wiggle
    // as the last wiggle
    vector<int> downs(n, 1);

    for (int i = 1; i < n; i++) {

      // if current nums is greater than previous num, that means it's a positive wiggle
      if (nums[i] > nums[i - 1]) {
        // increase the downs before the subsequence by 1
        ups[i] = downs[i - 1] + 1;
        downs[i] = downs[i - 1];
      }

      else if (nums[i] < nums[i - 1]) {
        // increase the ups before the subsequence by 1
        ups[i] = ups[i - 1];
        downs[i] = ups[i - 1] + 1;
      }

      else {
        ups[i] = ups[i - 1];
        downs[i] = downs[i - 1];
      }
    }

    return max(ups[n - 1], downs[n - 1]);
  }
};

File: leetcode/medium/376-wiggle-sub/lin-dp.py
"""
  @author Vortexx2
  Problem 376 - Wiggle Subsequence

  Runtime - 51 ms O(n)
  Memory Usage - 13.9 MB O(n)
"""
from typing import List


class Solution:

  def wiggleMaxLength(self, nums: List[int]) -> int:

    n = len(nums)
    ups = [1] * n
    downs = [1] * n

    for i in range(1, n):

      if nums[i] > nums[i - 1]:
        ups[i] = downs[i - 1] + 1
        downs[i] = downs[i - 1]

      elif nums[i] < nums[i - 1]:
        ups[i] = ups[i - 1]
        downs[i] = ups[i - 1] + 1

      else:
        ups[i] = ups[i - 1]
        downs[i] = downs[i - 1]

    return max(ups[n - 1], downs[n - 1])
